### PR TITLE
esp32: interrupt_controller: add support for interrupt allocation

### PR DESCRIPTION
**Pull request by glaubermaroto**
Adds interrupt allocation support for the ESP32 SoC. This implementation is largely based on ESP-IDF's v4.3 interrupt allocator, which allows dynamic interrupt lines assignment to the supported peripherals while keeping records of the lines already in use.

Additionally, the following drivers were updated to make use of this allocator:

* Counter
* Watchdog
* UART
* GPIO
* I2C
* SPI

### CODE REVIEW

**Comment by glaubermaroto** [comment]
Rebased to the main branch and ready for review.

@sylvioalves Please, take a look.

File: drivers/interrupt_controller/CMakeLists.txt [hunk 1]
```diff
@@ -20,3 +20,4 @@ zephyr_sources_ifdef(CONFIG_SWERV_PIC               intc_swerv_pic.c)
 zephyr_sources_ifdef(CONFIG_NPCX_MIWU               intc_miwu.c)
 zephyr_sources_ifdef(CONFIG_LEON_IRQMP              intc_irqmp.c)
 zephyr_sources_ifdef(CONFIG_INTEL_VTD_ICTL          intc_intel_vtd.c)
+zephyr_sources_ifdef(CONFIG_SOC_ESP32		    intc_esp32.c)
```
Comment (sylvioalves) [inline drivers/interrupt_controller/CMakeLists.txt:23]: Please, fix indentation to meet previous content alignment.

**Review by mbolivar-nordic — APPROVED** [review]
I don't know the arch, but the DT bits seem reasonable and it's nice to see these moving out of Kconfig.

**Review by sylvioalves — APPROVED** [review]

**Comment by sylvioalves** [comment]
> I don't know the arch, but the DT bits seem reasonable and it's nice to see these moving out of Kconfig.

Yes, we have been working that out to meet all zephyr's requirements.

**Pull request merged**